### PR TITLE
Call notify_get_state on non main thread

### PR DESCRIPTION
**Pull request by pvollan**
#### 2349fc0c0311c13a445dec02efeac6c3671a4692
<pre>
Call notify_get_state on non main thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=271827">https://bugs.webkit.org/show_bug.cgi?id=271827</a>
<a href="https://rdar.apple.com/125521998">rdar://125521998</a>

Reviewed by Chris Dumez.

To avoid blocking the main thread in the UI process, call notify_get_state on a non main thread.

* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerNotificationObservers):

Canonical link: <a href="https://commits.webkit.org/276841@main">https://commits.webkit.org/276841@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/48715219108eb5a61742125610c17872f259b393

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45854 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24982 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48438 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48525 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41891 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29276 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22380 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37525 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46432 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22053 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39550 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18700 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19453 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3898 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42152 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40995 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50293 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20848 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17344 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44653 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22150 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43532 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22509 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6393 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21841 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm [hunk 1]
```diff
@@ -726,17 +726,20 @@ static void logProcessPoolState(const WebProcessPool& pool)
...
+                if (!protectedThis)
+                    return;
+                if (!protectedThis->m_processes.isEmpty()) {
+                    String messageString(message);
+                    for (auto& process : protectedThis->m_processes)
+                        process->send(Messages::WebProcess::PostNotification(messageString, (status == NOTIFY_STATUS_OK) ? std::optional<uint64_t>(state) : std::nullopt), 0);
+                }
+            });
         });
         if (status)
```
Comment (cdumez) [inline Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm]: This patch seems wrong to me. I don't see how you can check `status` here synchronously on the main thread when it now gets set off the main thread, asynchronously.

**Review by cdumez — CHANGES_REQUESTED** [review]

File: Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm (hunk 1 above)
Comment (cdumez) [reply Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm]: Ohhh, there are 2 `status` variables, in different scopes. That's bad practice and why I got confused.

File: Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm [hunk 2]
```diff
@@ -726,17 +726,20 @@ static void logProcessPoolState(const WebProcessPool& pool)
...
-            }
+        auto queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
+        auto status = notify_register_dispatch(message, &notifyToken, queue, [weakThis, message](int token) {
+            uint64_t state = 0;
+            auto status = notify_get_state(token, &state);
+            callOnMainRunLoop([weakThis, message, state, status] {
+                RefPtr protectedThis = weakThis.get();
+                if (!protectedThis)
+                    return;
+                if (!protectedThis->m_processes.isEmpty()) {
```
Comment (cdumez) [inline Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm]: This should use `WebProcessPool::sendToAllProcesses()`.

File: Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm (hunk 1 above)
Comment (cdumez) [reply Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm]: So to clarify, I think this is correct but you should rename the variables so they don't have the same name.

**Review by cdumez — CHANGES_REQUESTED** [review]

File: Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm (hunk 1 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm]: Ah, that is a good point! I will address this in the upcoming patch.

Thanks for reviewing!

**Review by cdumez — APPROVED** [review]
LGTM

**Comment by pvollan** [comment]
> LGTM

Thanks for reviewing!

**Pull request merged**